src/llmq/ehf_signals.cpp

```diff
@@ -110,33 +117,37 @@ void CEHFSignalsHandler::HandleNewRecoveredSig(const CRecoveredSig& recoveredSig
         return;
     }
 
+    const auto ehfSignals = WITH_LOCK(cs_main, return chainstate.GetMNHFSignalsStage(chainstate.m_chain.Tip()));
```

Comment:
maybe same here; m_chain.Tip may need it, but doesn't seem to

Reply:
m_chain.Tip() should need cs_main locked since another thread could modify the same internal variable accessed at the same time (?)

Reply:
@UdjinM6 / @knst please double check me, but I don't think accessing .Tip() needs cs_main? It feels like it should need it, but I see no "EXCLUSIVE_LOCKS_REQUIRED" annotations

Reply:
```    /** Returns the index entry for the tip of this chain, or nullptr if none. */
    CBlockIndex *Tip() const {
        return vChain.size() > 0 ? vChain[vChain.size() - 1] : nullptr;
    }
```
but:
```
/**      
 * CChain implementation
 */
void CChain::SetTip(CBlockIndex *pindex) {
    if (pindex == nullptr) {
...
    }
    vChain.resize(pindex->nHeight + 1);
...
}
```

And `SetTip` is always called under `cs_main` mutex.

You can have some chain (that is not ActiveChain() and chain it somehow without any locks; but once you work with ActiveChain (and m_chain is it) - you need to call cs_main.

There's no annotation because there can exist multiple chain on same time and only one (which is changed in validation) should be guarded by cs_main

`cs_main` is needed here.

Reply:
can be updated though:
```
-    const auto ehfSignals = WITH_LOCK(cs_main, return chainstate.GetMNHFSignalsStage(chainstate.m_chain.Tip()));
+    const auto ehfSignals = chainstate.GetMNHFSignalsStage(WITH_LOCK(cs_main, return chainstate.m_chain.Tip()));
```

Reply:
review applied